### PR TITLE
[WIP]Implement written books

### PR DESCRIPTION
**Pull request by Seadragon91**
- Added new class cBookContent that contains the infos about a book
- Loads and saves the book from player inventory and nbt
- Added hooks: HOOK_PLAYER_EDITING_BOOK and HOOK_PLAYER_EDITED_BOOK

Revived that old branch. Tested and it works in 1.8.9, 1.9.4 and 1.12.2.
Will check the API and look to extend it with `cJson`.

### CODE REVIEW

**Comment by Seadragon91** [comment]
Had thought about adding a static function to deserialize from json string, but I think the API functions are already enough.

File: src/BookContent.h [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+	/** Clears the whole book */
+	void Clear();
+
+	/** Returns true if the book has no author, no title and no pages */
+	bool IsEmpty(void) const;
+
+	// tolua_end
+
+	/** Required in ManualBindings to save the page as simple string or json string */
```
Comment (bearbin) [inline src/BookContent.h:41]: I can't make any sense out of this comment, could you explain a bit more?

File: src/Protocol/Protocol_1_9.cpp [hunk 2]
```diff
@@ -3103,6 +3114,64 @@ void cProtocol_1_9_0::HandleVanillaPluginMessage(cByteBuffer & a_ByteBuffer, con
...
+		}
+		BookItem.m_BookContent.SetIsSigned(IsSigned);
+
+		// The equipped item contains the old book content
+		if (cRoot::Get()->GetPluginManager()->CallHookPlayerEditingBook(Player, Player.GetEquippedItem().m_BookContent, BookItem.m_BookContent, IsSigned))
+		{
+			// Plugin denied the player to edit the book
+			cInventory & inv = Player.GetInventory();
+			inv.SetHotbarSlot(inv.GetEquippedSlotNum(), BookItem);
+			Player.GetInventory().SendEquippedSlot();
```
Comment (bearbin) [inline src/Protocol/Protocol_1_9.cpp:3163]: The plugin denied editing the book but we set it anyway? This looks wrong.

File: src/Protocol/Protocol_1_9.cpp [hunk 3]
```diff
@@ -3103,6 +3114,64 @@ void cProtocol_1_9_0::HandleVanillaPluginMessage(cByteBuffer & a_ByteBuffer, con
...
+			return;
+		}
+
+		cInventory & inv = Player.GetInventory();
+		inv.SetHotbarSlot(inv.GetEquippedSlotNum(), BookItem);
+		Player.GetInventory().SendEquippedSlot();
```
Comment (bearbin) [inline src/Protocol/Protocol_1_9.cpp:3169]:
```suggestion
		Player.GetInventory().SetEquippedItem(BookItem);
```

File: src/Protocol/Protocol_1_8.cpp [hunk 4]
```diff
@@ -2969,6 +2980,65 @@ void cProtocol_1_8_0::HandleVanillaPluginMessage(cByteBuffer & a_ByteBuffer, con
...
+			// Plugin denied the player to edit the book
+			cInventory & inv = Player.GetInventory();
+			inv.SetHotbarSlot(inv.GetEquippedSlotNum(), BookItem);
+			Player.GetInventory().SendEquippedSlot();
+			return;
+		}
+
+		cInventory & inv = Player.GetInventory();
+		inv.SetHotbarSlot(inv.GetEquippedSlotNum(), BookItem);
+		Player.GetInventory().SendEquippedSlot();
```
Comment (bearbin) [inline src/Protocol/Protocol_1_8.cpp:3036]:
```suggestion
		Player.GetInventory().SetEquippedItem(BookItem);
```

File: src/Protocol/Protocol_1_8.cpp [hunk 5]
```diff
@@ -2969,6 +2980,65 @@ void cProtocol_1_8_0::HandleVanillaPluginMessage(cByteBuffer & a_ByteBuffer, con
...
+			IsSigned = false;
+			BookItem = cItem(E_ITEM_BOOK_AND_QUILL);
+			cBookContent::ParseFromNBT(0, BookItem.m_BookContent, NBT);
+		}
+		BookItem.m_BookContent.SetIsSigned(IsSigned);
+
+		// The equipped item contains the old book content
+		if (cRoot::Get()->GetPluginManager()->CallHookPlayerEditingBook(Player, Player.GetEquippedItem().m_BookContent, BookItem.m_BookContent, IsSigned))
+		{
+			// Plugin denied the player to edit the book
```
Comment (bearbin) [inline src/Protocol/Protocol_1_8.cpp:3027]: Again, plugin denied editing but we do it anyway?

Comment (Seadragon91) [reply src/Protocol/Protocol_1_8.cpp:3027]: Fixed, thanks.

File: src/Protocol/Protocol_1_8.cpp (hunk 4 above)
Comment (Seadragon91) [reply src/Protocol/Protocol_1_8.cpp]: Fixed, thanks.

File: src/BookContent.h (hunk 1 above)
Comment (Seadragon91) [reply src/BookContent.h]: Changed the comment. In a written book the content is a json string and in book and quill a simple string.

**Comment by Seadragon91** [comment]
I tested against 1.8 and there written book are stored as simple strings and not as json strings.

**Comment by Seadragon91** [comment]
Okay, it' sent as a json string in 1.8, but the text is escaped:
1.9:
```json
{
   "text" : "1\n2\n3\n4\n5"
}
```

1.8:
```json
{
   "text" : "\"1\\n2\\n3\\n4\\n5\""
}
```
Should we change the content, before it's sent? Or ignore it as it's only in 1.8?

**Comment by bearbin** [comment]
When it's received seems to make more sense. Since we do support 1.8 still, we should make all features that can work in that version work.

**Pull request closed without merging**